[Zen2] Update documentation for Zen2 (#34714)
elastic/elasticsearch: docs/reference/modules/discovery/adding-removing-nodes.asciidoc

[[modules-discovery-adding-removing-nodes]]
=== Adding and removing nodes

> 

Reviewer: It seems like the majority of this information pertains only to removing nodes.   It's not necessary immediately, but at some point I think it would be good to split this into two separate pages -- one about adding nodes (with lots of details about how that differs depending on platform and node type, etc) and one about removing nodes (which would be most of this content).

As nodes are added or removed Elasticsearch maintains an optimal level of fault
tolerance by automatically updating the cluster's _voting configuration_, which

> is the set of master-eligible nodes whose responses are counted when making

Reviewer: ```suggestion
is the set of <<master-node,master-eligible nodes>> whose responses are counted when making
```

master-ineligible nodes only. However there are situations in which it may be
desirable to add or remove some master-eligible nodes to or from a cluster.

If you wish to add some master-eligible nodes to your cluster, simply configure
the new nodes to find the existing cluster and start them up. Elasticsearch will
add the new nodes to the voting configuration if it is appropriate to do so.

When removing master-eligible nodes, it is important not to remove too many all
at the same time. For instance, if there are currently seven master-eligible
nodes and you wish to reduce this to three, it is not possible simply to stop
four of the nodes at once: to do so would leave only three nodes remaining,
which is less than half of the voting configuration, which means the cluster
cannot take any further actions.

As long as there are at least three master-eligible nodes in the cluster, as a
general rule it is best to remove nodes one-at-a-time, allowing enough time for
the cluster to <<modules-discovery-quorums,auto-adjust>> the voting
configuration and adapt the fault tolerance level to the new set of nodes.

If there are only two master-eligible nodes remaining then neither node can be

> safely removed since both are required to reliably make progress, so you must

Reviewer: ```suggestion
safely removed since both are required to reliably make progress. You must
```

first inform Elasticsearch that one of the nodes should not be part of the
voting configuration, and that the voting power should instead be given to

> other nodes, allowing the excluded node to be taken offline without preventing

Reviewer: ```suggestion
other nodes. You can then take the excluded node offline without preventing
```

the other node from making progress. A node which is added to a voting

> configuration exclusion list still works normally, but Elasticsearch will try

Reviewer: ```suggestion
configuration exclusion list still works normally, but Elasticsearch
```

> and remove it from the voting configuration so its vote is no longer required.

Reviewer: ```suggestion
tries to remove it from the voting configuration so its vote is no longer required.
```

Importantly, Elasticsearch will never automatically move a node on the voting
exclusions list back into the voting configuration. Once an excluded node has
been successfully auto-reconfigured out of the voting configuration, it is safe
to shut it down without affecting the cluster's master-level availability. A
node can be added to the voting configuration exclusion list using the
following API:

[source,js]
--------------------------------------------------
# Add node to voting configuration exclusions list and wait for the system to
# auto-reconfigure the node out of the voting configuration up to the default
# timeout of 30 seconds
POST /_cluster/voting_config_exclusions/node_name

# Add node to voting configuration exclusions list and wait for
# auto-reconfiguration up to one minute
POST /_cluster/voting_config_exclusions/node_name?timeout=1m
--------------------------------------------------
// CONSOLE
// TEST[skip:this would break the test cluster if executed]

The node that should be added to the exclusions list is specified using
<<cluster-nodes,node filters>> in place of `node_name` here. If a call to the

> voting configuration exclusions API fails then the call can safely be retried.

Reviewer: ```suggestion
voting configuration exclusions API fails, you can safely retry it.
```

configuration exclusions list, which has the system automatically try to
reconfigure the voting configuration to remove that node and prevents it from
returning to the voting configuration once it has removed. The current list of
exclusions is stored in the cluster state and can be inspected as follows:

[source,js]
--------------------------------------------------
GET /_cluster/state?filter_path=metadata.cluster_coordination.voting_config_exclusions
--------------------------------------------------
// CONSOLE

This list is limited in size by the following setting:

`cluster.max_voting_config_exclusions`::

    Sets a limits on the number of voting configuration exclusions at any one
    time.  Defaults to `10`.

Since voting configuration exclusions are persistent and limited in number, they
must be cleaned up. Normally an exclusion is added when performing some
maintenance on the cluster, and the exclusions should be cleaned up when the
maintenance is complete. Clusters should have no voting configuration exclusions
in normal operation.

If a node is excluded from the voting configuration because it is to be shut

> down permanently then its exclusion can be removed once it has shut down and

Reviewer: ```suggestion
down permanently, its exclusion can be removed after it is shut down and
```

> been removed from the cluster. Exclusions can also be cleared if they were

Reviewer: ```suggestion
removed from the cluster. Exclusions can also be cleared if they were
```
